addon/services/prefetch.js

```diff
@@ -0,0 +1,45 @@
+import Service, { inject as service } from '@ember/service';
+import { schedule } from '@ember/runloop';
+import { createPrefetchChangeSet } from '../-private/diff-route-info';
+import RSVP from 'rsvp';
+import { gte } from 'ember-compatibility-helpers';
+
+let PrefetchService;
+
+if (gte('3.6.0')) {
+  PrefetchService = Service.extend({
+    router: service('router'),
+    init() {
+      this._super(...arguments);
+      let privateRouter = this.router._router._routerMicrolib;
+      let seenRoutes = new WeakMap();
+
+      this.router.on('routeWillChange', transition => {
+        if (transition.to && /(^|_|\.)(loading$|error$)/.test(transition.to.name)) {
+          return;
+        }
+
+        schedule('actions', () => {
+          let changeSet = createPrefetchChangeSet(privateRouter, transition);
```

Comment:
Should this schedule be guarded on if the service has been destroyed or not? I

Reply:
I actually think you want the body of this to check since it is the thing that is actually closing over the service. Good catch.